src/Common/src/Interop/Unix/libssl/SecuritySafeHandles.cs

```diff
@@ -265,12 +276,23 @@ public override bool IsInvalid
 
         protected override bool ReleaseHandle()
         {
-            Interop.OpenSsl.FreeSslContext(_sslContext);
+            Interop.OpenSsl.Disconnect(_sslContext);
```

Comment:
Disconnect is needed only if handshake has completed and this is a bug in our existing logic that we are not checking for this. By adding this state to SafeSslHandle (and setting it in DoSslHandshake), we should be able to move the Disconnect logic to SafeSslHandle.ReleaseHandle.